### PR TITLE
gRPC: Implement endpoint 'getAccountInfo'

### PR DESCRIPTION
**Pull request by bisgardo**
Added `getAccountInfo` to `Client` along with wrappers for all related generated types to keep it from being too unpleasant to work with.

Resolves https://concordium.atlassian.net/browse/CBW-1563.

### CODE REVIEW

**Comment by rh-concordium** [comment]
Sidenote: I THINK that if you provide a link to the jira in question in the PR description, then it establish the jira/Github integration
(disclaimer: Not 100% sure how the integration works, so it MIGHT be necessary to also mention the jira number in the branch name for the integration to work. With Bitbucket you only have to include the jira number in the branch name to make it
Bitbucket-jira
 work)

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (rh-concordium) [inline Sources/ConcordiumSwiftSDK/Model/Account.swift:300]: It is intentional that we haven't included the `unrecognized` case to this enum, like in the generated one? This method will throw an exception in case we come across the unrecognized value

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (rh-concordium) [inline Sources/ConcordiumSwiftSDK/Model/Account.swift:305]: Do we need to make these fields optional (like in Rust?) or have some kind of indication whether the fields have been explicitly set (like in the generated Swift type)

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (rh-concordium) [inline Sources/ConcordiumSwiftSDK/Model/Account.swift:363]: Should we add tests for the different cases?

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (rh-concordium) [inline Sources/ConcordiumSwiftSDK/Model/Account.swift:309]: The `fromGrpc` method is more or less ubiquitous, so we could consider adding a generic protocol for it. Then again, we would need to add two protocols - one for `fromGrpcType` methods that throws, and one that doesn't, so it MIGHT be overkill for our purposes

File: Tests/ConcordiumSwiftSDKTests/ClientTests.swift
Comment (rh-concordium) [inline Tests/ConcordiumSwiftSDKTests/ClientTests.swift:70]: Is there anything we can do about that (i.e., should we create a jira), or is there something that the Blockchain team should do about that, or will it never happen in real life.

**Review by rh-concordium — APPROVED** [review]
Overall, it looks very nice!
I only have a few (clarifying) questions on the extent to which it makes sense for us to add tests, and I noticed a few discrepancies to the Rust implementation (which might be intentional)

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:155]: We could. I prefer type params to have single-letter names to distinguish them from real types though. I think it's the same in the rust impl, though the params have restrictions that add some documentational value.

It's worth giving these types another iteration. I'm inclined to prefer doing that after merging this. Would it be possible for you to do that while I finish up formatting and start on the binding stuff?

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:162]: What if I reword it to "in the Rust SDK this is decoded into struct"?

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:165]: Will check. I might have used names from the Java SDK.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:171]: Indeed 👯

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:185]: Let's keep the discussion in the same thread above.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:247]: I'll consider this part of the "needs another iteration" discussion above.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:275]: Discussed above.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:305]: It's not, I'll give it another look.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:309]: What would be the advantage of that? Also, can you have static functions in protocols?

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:363]: Discussed above.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:389]: yes will do that

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:390]: Agree it's weird. Did it because the SDKs do (some of them at least). Should we rename?

File: Tests/ConcordiumSwiftSDKTests/ClientTests.swift
Comment (bisgardo) [reply Tests/ConcordiumSwiftSDKTests/ClientTests.swift:70]: I think it's an issue with the generated client. But haven't dug into it. I'll create a task.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:300]: Looks like an oversight. I'll give it another look.

File: Package.swift
Comment (rh-concordium) [reply Package.swift:9]: Agree, a comment would only be (potentially) useful in the case where we have a have a single dependency/reason that forces us to set a specific minimum version, so I am completely fine with leaving it out, so that the comment doesn't start lying the second we have additional reasons to bump the version further.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (rh-concordium) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:390]: Well, since the context (the type) is right there, we probably won't confuse anyone by removing the "account" prefix - but it's not something I feel strongly about, so I'll leave it up to you :)

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (rh-concordium) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:309]: Ha - I missed the static keyword - which is weird since it's a factory method :D

The only advantage in our case would be to ensure naming consistency.   

Swift protocols can define static members, BUT the static members can't refer to the protocol's associatedtype, so that settles it :)

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:386]: Agree I'll do that and similar cases in other types.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:303]: Fixed. I didn't notice this because of the retarded way the generated grpc types work. Nice catch!

File: Tests/ConcordiumSwiftSDKTests/ClientTests.swift
Comment (bisgardo) [reply Tests/ConcordiumSwiftSDKTests/ClientTests.swift:70]: Filed https://concordium.atlassian.net/browse/CBW-1641. Will remove the TODO.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:232]: I gave it another iteration, adding docstrings from Rust SDK and compared with Java SDK as well. I think it's fine for now.

**Comment by bisgardo** [comment]
> Sidenote: I THINK that if you provide a link to the jira in question in the PR description, then it establish the jira/Github integration (disclaimer: Not 100% sure how the integration works, so it MIGHT be necessary to also mention the jira number in the branch name for the integration to work. With Bitbucket you only have to include the jira number in the branch name to make it Bitbucket-jira work)

It looks like adding a link in the description indeed did the trick 👍

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift]: Updated.

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:206]: On a closer look it seems like the Rust SDK intentially reduces the type size of some values that are defined as 32 bit in the protobuf spec. @abizjak can you shed some light on this?

Comment (abizjak) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:206]: The real type is this https://github.com/Concordium/concordium-base/blob/main/rust-src/concordium_base/src/id/types.rs#L379

Protobuf does not have u16 and u8 so we do what we can use there. But the values will always fit the range stated there.

Comment (abizjak) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:206]: Further to this, the value for year is actually meant to be in rane 1000..=9999 and month in the obvious range 1..=12.

Also note that a JSON format is based on a string in the format YYYYMM (this will probably become relevant when you do FFI).

Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:206]: I think for the purpose of this SDK it's fine to keep the types as u32. I think that's what the Java SDK does as well. Do you agree @abizjak ? The ranges are already documented in the docstring. I'll add a method for the JSON format.

Comment (abizjak) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:206]: Yes.

Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift:206]: done

File: Sources/ConcordiumSwiftSDK/Model/Account.swift
Comment (bisgardo) [reply Sources/ConcordiumSwiftSDK/Model/Account.swift]: Let me reformulate this: Such tests would be useful but very cumbersome to write. The transformation is almost trivial and the type system helps a lot. The most likely problem is that we make fields that are not nillable but should be. But the structs are based on the ones from the Rust SDK so I think that's unlikely too.

You're right about the error cases. Error reporting is very bad atm (and thus also not worth testing in it's current state); there's a task for improving that.

**Review by rh-concordium — APPROVED** [review]
Looks good +1

**Pull request merged**
